### PR TITLE
Fix false positive for `unused-argument` where nested function uses parent argument as a `nonlocal`

### LINKED ISSUE #5187: Pylint doesn't see nonlocal usage of variable
Basically similar to #2671

### Bug description

```python

def myfunction(function, arg1, arg2):
    """ lorem ipsum """

    def decorator():
        """ lorem ipsum """
        nonlocal arg2
        arg2 += 1

        return function(arg1, arg2)

    return decorator

```

Error: 
`W0613: Unused argument 'arg2' (unused-argument)`

### Configuration

_No response_

### Command used

```shell
pylint src/lorem.py
```

### Pylint output

```shell
************* Module lorem
src/lorem.py:3:31: W0613: Unused argument 'arg2' (unused-argument)

------------------------------------------------------------------
Your code has been rated at 8.33/10 (previous run: 8.33/10, +0.00)
```

### Expected behavior

Detect nonlocal usage of arg2

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.2
Python 3.7.12 (default, Sep 10 2021, 17:03:41) [Clang 11.0.0 (clang-1100.0.33.17)]
```

### OS / Environment

MacOS 10.14.6

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

When returning a nested function without calling it, and when the returned function had a `nonlocal` name coinciding with one of the parent function's arguments, that argument was inappropriately marked as unused on the parent.

Closes #5187

### CODE REVIEW

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -1295,8 +1296,12 @@ def leave_functiondef(self, node: nodes.FunctionDef) -> None:
         if is_method and node.is_abstract():
             return
 
-        global_names = _flattened_scope_names(node.nodes_of_class(nodes.Global))
-        nonlocal_names = _flattened_scope_names(node.nodes_of_class(nodes.Nonlocal))
+        global_names: Set[str] = _flattened_scope_names(
```
Comment (DanielNoord) [inline pylint/checkers/variables.py:1299]:
```diff
diff --git a/pylint/checkers/variables.py b/pylint/checkers/variables.py
index 1781866aa..7f36201e3 100644
--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -67,6 +67,7 @@ from typing import (
     Any,
     DefaultDict,
     Dict,
+    Iterator,
     List,
     NamedTuple,
     Optional,
@@ -372,7 +373,9 @@ def _import_name_is_global(stmt, global_names):
     return False
 
 
-def _flattened_scope_names(iterator):
+def _flattened_scope_names(
+    iterator: Iterator[Union[nodes.Global, nodes.Nonlocal]]
+) -> Set[str]:
     values = (set(stmt.names) for stmt in iterator)
     return set(itertools.chain.from_iterable(values))
 ```

This should make these unnecessary :)

**Review by DanielNoord — CHANGES_REQUESTED** [review]

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
LGTM.  It needs a rebase and the changelog need to be for 2.14 :)

File: ChangeLog [hunk 2]
```diff
@@ -10,6 +10,10 @@ Release date: TBA
   Put new features here and also in 'doc/whatsnew/2.14.rst'
 
 
+* Fixed false positive for ``unused-argument`` when a ``nonlocal`` name is used
```
Comment (DanielNoord) [inline ChangeLog:13]: Shouldn't this target `2.13.1`? @Pierre-Sassoulas

Comment (Pierre-Sassoulas) [reply ChangeLog:13]: How about we backport all bug fixes and false positives in last minor, keeping only features and false negatives in the next minor/major ?

**Review by Pierre-Sassoulas — APPROVED** [review]

File: ChangeLog (hunk 2 above)
Comment (DanielNoord) [reply ChangeLog:13]: I think you meant "in last patch" (_Sem ver names are awful_), but you I agree!

Comment (Pierre-Sassoulas) [reply ChangeLog:13]: In my head it's always "breaking change", "major", "minor" instead of "major", "minor" "patch" 😄

**Pull request merged**